subprojects/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/CachingJvmMetadataDetector.java

```diff
@@ -47,4 +49,10 @@ private File resolveSymlink(File jdkPath) {
         }
     }
 
+    @Override
+    public void invalidateItemsMatching(Predicate<JvmInstallationMetadata> predicate) {
+        synchronized (javaMetadata) {
```

Comment:
Maybe use a synchronized map instead of a concurrent one, to make sure this really works as intended all the time.

Reply:
What are you exactly trying to prevent with the `synchronized` block?
It will only help if invalidation is called concurrently, but does not protect about adding entries concurrently.